mobile/src/main/java/ch/epfl/sdp/mobile/data/api/firebase/FirebaseAuthenticationApi.kt

```diff
@@ -0,0 +1,86 @@
+package ch.epfl.sdp.mobile.data.api.firebase
+
+import ch.epfl.sdp.mobile.data.api.AuthenticationApi
+import ch.epfl.sdp.mobile.data.api.AuthenticationApi.AuthenticationResult
+import ch.epfl.sdp.mobile.data.api.AuthenticationApi.AuthenticationResult.Failure
+import ch.epfl.sdp.mobile.data.api.AuthenticationApi.AuthenticationResult.Success
+import com.google.firebase.auth.AuthResult
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.auth.FirebaseUser
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.tasks.await
+
+/**
+ * An implementation of the [AuthenticationApi] which uses [FirebaseAuth] internally.
+ *
+ * @param auth the [FirebaseAuth] instance which will be used to handle authentication.
+ */
+class FirebaseAuthenticationApi(private val auth: FirebaseAuth) : AuthenticationApi {
+
+  override val currentUser: Flow<AuthenticationApi.User> =
+      auth.currentUserFlow()
+          .map { user ->
```

Comment:
[Proposition] It took me some time to understand what was going on here (maybe because I am new to Kotlin) However I suggest this 😃.  What do you think ?  
```suggestion
          .mapNotNull { user ->  FirebaseUserAdapter(it, auth)}
          .onEmpty { AuthenticationApi.User.NotAuthenticated }
          .onStart { emit(AuthenticationApi.User.Loading) }
```

Reply:
I think the semantics of this suggested change are slightly different than what I had in mind - in particular, the `mapNotNull` operator will filter out `null` values in the resulting Flow, when we want to preserve them. Also, `onEmpty` will simply let the `Flow` emit certain values if it's empty, but won't be used otherwise. It's a good way to provide a "default value" to Flows which might emit 0 elements.

Just fyi, when using an operator like `onEmpty {}`, you should be cautious to call `emit` if you want to emit an element; otherwise, the statement is executed but won't issue items in the Flow.

Your concerns about the readability of the code are valid though. I'll extract this code to a dedicated function to make it easier to grasp.

Reply:
Done in commit 17000705893bc72125209f43936ff838d265c276 😄